### ISSUE TITLE
Animation is smooth at first but flickers when the mouse hovers on canvas

### ISSUE THREAD

**Issue by Zaid-Ajaj** [MangelMaxime/Elmish.Canvas#1]
Hello there @MangelMaxime, very nice library you made, I wanted to play with a canvas in elmish for a long time.  
Only I noticed that the animation of the sample becomes sluggish and starts to flicker once the mouse enters the canvas, if the mouse didn't enter the canvas, the default animation stays smooth. Did you notice that as well? I see this behavior both in Chrome and Edge

**Comment by MangelMaxime** [comment]
I just publish the demo so I can test one of the components and docs features.

I notice some slow animation in the past but now seems better to me. But this is definitely something I want to take look at before people can play with. 
I want to use RAF instead of a promise so the animation will be easier to make smooth as the browser will ask for the frame only when ready and provide us the accurate delta time.

Here by using the promise to Tick the animation we don't have this precision.

Last thing, if you see it flickering when moving the mouse it's probably because the mouse generate a lot of events and for each of them we update the model. This is also something I want to work on for this app and any Elmish application. I need to create a debouncer/throtlle usable in the "view's dispatch" function so we can control how many event we handle in the app. :)

Thanks for taking a look :)

**Comment by MangelMaxime** [comment]
Oups, I forgot that I am already using `RAF` to trigger the `Tick` message :)

So the flickering is probaly due to the `Mouse` event. And I also, need for more complex example to include the timestamp delta to smooth the update.

**Comment by MangelMaxime** [comment]
Ok this freezing behavior seems to due to nested RAF one from Elmish and one from the react components. Using `withReactUnoptimized` seems to fix the issue and chrome FPS counter stay at 60fps all the time.

Firefox, move between 30-60fps.

**Comment by Zaid-Ajaj** [comment]
Thanks a lot for the explanation Maxime, just tried the example and it runs smoothly this time, keep up the good work :smile:

By the way, are going to make a library from this example?

**Comment by MangelMaxime** [comment]
Yes, I will make a library  so people can re-use the canvas thing and have the canvas components already setup. 

Before that, I have some things to look at like:
- performance
- make the HMR works if possible
  - The code is being updated but because there is an old frame still processed then the UI is restoring is old state. I had a similar problem for Thoth.Toast I think, and solved it so should be doable :)

PS: I am also using this issue to track my progress so I don't forget what I tried or how I fixed the performance issue :)

**Comment by MangelMaxime** [comment]
Allow smother animation and inputs handling.

In order to have smoother animation, we should take inspiration from game rendering.
- Weight current delta/fps compare to old one
  - Track fps on a preriod of time (like 10 sec) to have an accurate FPS count and not an estimate
- Simulate multiple ticks if the framerate drop
- Use a fixed delta for simulation tick
- Detect death spiral loop
  - If the framerate goes to low, we will try to simulate multiple ticks per frame but this take time and lower the framerate. So we have more ticks to simulate after, etc.
  - Detecting this behavior allow us to enter in `panic` which allow us to lower the resolution, rendering process, etc.
- Provide a `interp` value, so we can interpolate forward the animation
- Allow to play/pause/stop the frame requests

We should also try to provide the inputs register from the canvas components. And allow the user to set the `throtle` delay. I guess by default we could make it to `1second/fps` delta. More event handling means we are going to stress more the Elmish application and so create more Garbage collector pass.

Try to integrate into the canvas component the detection of HMR and cancel previous frames. (try preventing the flickering state)

**Comment by Zaid-Ajaj** [comment]
Thanks for explanation (and sorry for late answer, I have been on vacation) looks like a steady progress, although I think you can publish an alpha/experiemental version as is (maybe also a repo that acts a template to easily get started) and optimize later for the stable release, this way other people can play with it too :smile:
